src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,408 @@
+                        case AndSequence seq:
+                            var testsToInsert = seq.Nodes;
+                            tests.RemoveAt(i);
+                            for (int j = 0, n = testsToInsert.Length; j < n; j++)
+                                tests.Insert(i + j, testsToInsert[j]);
+                            break;
+                    }
+                }
+                var result = tests.Count switch
+                {
+                    0 => True.Instance,
+                    1 => tests[0],
+                    _ => new AndSequence(tests.ToImmutable()),
+                };
+                tests.Free();
+                return result;
+            }
+        }
+
+        internal sealed class OrSequence : Sequence
+        {
+            public OrSequence(ImmutableArray<AnalyzedNode> nodes) : base(nodes) { }
+
+            public override AnalyzedNode Update(ArrayBuilder<AnalyzedNode> nodes) => Create(nodes);
+            public override AnalyzedNode Negate(ArrayBuilder<AnalyzedNode> nodes) => AndSequence.Create(nodes);
+
+            public static AnalyzedNode Create(ArrayBuilder<AnalyzedNode> tests)
+            {
+                for (var i = tests.Count - 1; i >= 0; i--)
+                {
+                    switch (tests[i])
+                    {
+                        case False _:
+                            tests.RemoveAt(i);
+                            break;
+                        case True t:
+                            tests.Free();
+                            return t;
+                        case OrSequence seq:
+                            tests.RemoveAt(i);
+                            var testsToInsert = seq.Nodes;
+                            for (int j = 0, n = testsToInsert.Length; j < n; j++)
+                                tests.Insert(i + j, testsToInsert[j]);
+                            break;
+                    }
+                }
+                var result = tests.Count switch
+                {
+                    0 => False.Instance,
+                    1 => tests[0],
+                    _ => new OrSequence(tests.ToImmutable()),
+                };
+                tests.Free();
+                return result;
+            }
+        }
+
+        internal sealed class Pair : AnalyzedNode
+        {
+            public readonly Evaluation Input;
+            public readonly AnalyzedNode Pattern;
+
+            public Pair(Evaluation input, AnalyzedNode pattern)
+            {
+                Input = input;
+                Pattern = pattern;
+            }
+        }
+
+        internal abstract class Test : AnalyzedNode
+        {
+            public readonly Evaluation? Input;
+            protected Test(Evaluation? input) => Input = input;
+            public Test WithInput(Evaluation? newInput) => ReferenceEquals(newInput, Input) ? this : WithInputCore(newInput);
+            protected abstract Test WithInputCore(Evaluation? newInput);
```

Comment:
your desire for horizontal compatness is not really keeping style with the codebase.  it's fine for the above to be: